// @mangrovedao/mangrove-core/index.js
// DO NOT MODIFY -- GENERATED BY buildIndex.js
exports.abis = {};
exports.abis['AaveDeepRouter'] = require("./dist/mangrove-abis/AaveDeepRouter.json");
exports.abis['AaveV3Module'] = require("./dist/mangrove-abis/AaveV3Module.json");
exports.abis['AbstractRouter'] = require("./dist/mangrove-abis/AbstractRouter.json");
exports.abis['AccessControlled'] = require("./dist/mangrove-abis/AccessControlled.json");
exports.abis['ICreditDelegationToken'] = require("./dist/mangrove-abis/ICreditDelegationToken.json");
exports.abis['IERC20'] = require("./dist/mangrove-abis/IERC20.json");
exports.abis['ILiquidityProvider'] = require("./dist/mangrove-abis/ILiquidityProvider.json");
exports.abis['Mangrove'] = require("./dist/mangrove-abis/Mangrove.json");
exports.abis['MangroveOrder'] = require("./dist/mangrove-abis/MangroveOrder.json");
exports.abis['MangroveOrderEnriched'] = require("./dist/mangrove-abis/MangroveOrderEnriched.json");
exports.abis['MgvCleaner'] = require("./dist/mangrove-abis/MgvCleaner.json");
exports.abis['MgvOracle'] = require("./dist/mangrove-abis/MgvOracle.json");
exports.abis['MgvReader'] = require("./dist/mangrove-abis/MgvReader.json");
exports.abis['OfferMaker'] = require("./dist/mangrove-abis/OfferMaker.json");
exports.abis['SimpleTestMaker'] = require("./dist/mangrove-abis/SimpleTestMaker.json");
exports.abis['TestToken'] = require("./dist/mangrove-abis/TestToken.json");
exports.addresses = {};
exports.addresses.deployed = {};
exports.addresses.deployed['local'] = require("./addresses/deployed/local.json");
exports.addresses.deployed['matic'] = require("./addresses/deployed/matic.json");
exports.addresses.deployed['maticmum'] = require("./addresses/deployed/maticmum.json");
exports.addresses.context = {};
exports.addresses.context['matic'] = require("./addresses/context/matic.json");
exports.addresses.context['maticmum'] = require("./addresses/context/maticmum.json");
