// @mangrovedao/mangrove.js/dist/nodejs/aaveV3Module.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AaveV3Module_instances, _AaveV3Module_debtToken;
Object.defineProperty(exports, "__esModule", { value: true });
const ethers = __importStar(require("ethers"));
const types_1 = require("./types");
/**
 * The OfferLogic class connects to a OfferLogic contract.
 * It posts onchain offers.
 */
// OfferLogic.withdrawDeposit()
// OfferLogic.deposit(n)
class AaveV3Module {
    constructor(mgv, address, signer) {
        _AaveV3Module_instances.add(this);
        this.mgv = mgv;
        this.contract = types_1.typechain.AaveV3Module__factory.connect(address, signer ? signer : this.mgv._signer);
    }
    async approveDelegation(tokenName, borrower, overrides = {}) {
        const dTtkn = await __classPrivateFieldGet(this, _AaveV3Module_instances, "m", _AaveV3Module_debtToken).call(this, tokenName);
        return dTtkn.approveDelegation(borrower, ethers.constants.MaxUint256, overrides);
    }
    async status(tokenName, account) {
        const asset = this.mgv.token(tokenName);
        const dToken = await __classPrivateFieldGet(this, _AaveV3Module_instances, "m", _AaveV3Module_debtToken).call(this, tokenName);
        const { maxRedeemableUnderlying, maxBorrowAfterRedeemInUnderlying } = await this.contract.maxGettableUnderlying(asset.address, true, account);
        return {
            available: asset.fromUnits(maxRedeemableUnderlying),
            borrowable: asset.fromUnits(maxBorrowAfterRedeemInUnderlying),
            borrowing: asset.fromUnits(await dToken.balanceOf(account)),
        };
    }
    async logStatus(tokenNames, account) {
        account = account ? account : await this.mgv._signer.getAddress();
        for (const tokenName of tokenNames) {
            const stat = await this.status(tokenName, account);
            console.log(`----------${tokenName}----------`);
            console.log("debit:", `\u001b[32m${stat.available}\u001b[0m`);
            console.log("credit:", `\u001b[33m${stat.borrowable}\u001b[0m`);
            console.log("debt:", `\u001b[31m${stat.borrowing}\u001b[0m`);
        }
    }
}
_AaveV3Module_instances = new WeakSet(), _AaveV3Module_debtToken = async function _AaveV3Module_debtToken(tokenName, signer) {
    const asset_address = this.mgv.token(tokenName).address;
    const debt_address = await this.contract.debtToken(asset_address);
    return types_1.typechain.ICreditDelegationToken__factory.connect(debt_address, signer ? signer : this.mgv._signer);
};
exports.default = AaveV3Module;
