// @mangrovedao/mangrove.js/dist/nodejs/index.js
"use strict";
/**
 * @file Mangrove
 * @desc This file defines the exports of the `mangrove.js` package.
 * @hidden
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestMaker = exports.mgvTestUtil = exports.LiquidityProvider = exports.OfferLogic = exports.MgvToken = exports.Semibook = exports.Market = exports.Mangrove = exports.ethers = exports.decimals = exports.typechain = exports.eth = void 0;
const ethers_1 = require("ethers");
Object.defineProperty(exports, "ethers", { enumerable: true, get: function () { return ethers_1.ethers; } });
const eth = __importStar(require("./eth"));
exports.eth = eth;
const constants_1 = require("./constants");
Object.defineProperty(exports, "decimals", { enumerable: true, get: function () { return constants_1.decimals; } });
const mangrove_1 = __importDefault(require("./mangrove"));
exports.Mangrove = mangrove_1.default;
const market_1 = __importDefault(require("./market"));
exports.Market = market_1.default;
const semibook_1 = __importDefault(require("./semibook"));
exports.Semibook = semibook_1.default;
const offerLogic_1 = __importDefault(require("./offerLogic"));
exports.OfferLogic = offerLogic_1.default;
const mgvtoken_1 = __importDefault(require("./mgvtoken"));
exports.MgvToken = mgvtoken_1.default;
const liquidityProvider_1 = __importDefault(require("./liquidityProvider"));
exports.LiquidityProvider = liquidityProvider_1.default;
const TestMaker_1 = __importDefault(require("./util/test/TestMaker"));
exports.TestMaker = TestMaker_1.default;
const mgvTestUtil = __importStar(require("./util/test/mgvIntegrationTestUtil"));
exports.mgvTestUtil = mgvTestUtil;
const types_1 = require("./types");
Object.defineProperty(exports, "typechain", { enumerable: true, get: function () { return types_1.typechain; } });
// Turn off Ethers.js warnings
// ethers.utils.Logger.setLogLevel(ethers.utils.Logger.levels.ERROR);
exports.default = mangrove_1.default;
