// @mangrovedao/mangrove.js/dist/nodejs/util/runTypechain.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
Generate typechain files from mangrove-core ABIs/artifacts.
*/
const shelljs_1 = __importDefault(require("shelljs"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const typechain_1 = require("typechain");
/* This script gathers all relevant .json artifacts and runs typechain on them.
    Since the .json file come from multiple places, it is not as easy as doing input -> typechain -> outpout
  2 methods were tried but did not work.
    1) Running typechains with artifacts = glob(cwd,[dir1/*.json,dir2/*.json]) will generate .ts files with their directory hierarchy up to the lowest common ancestor of all globbed files. We want to flatten the file hierarchies so this does not work.
    2) Running typechain twice (ie. runTypeChain(...); runTypeChain(...);) will fail because the second invocation will erase the typechain/index.ts file generated at the first invocation.
  The current method is to copy everything to a temp directory, then run typechain.
*/
async function main() {
    // Clean out directory
    const outDir = path_1.default.join(process.cwd(), "src/types/typechain");
    shelljs_1.default.rm("-rf", outDir);
    shelljs_1.default.mkdir("-p", outDir);
    const pwd = shelljs_1.default.pwd();
    // Generate temp directory to place artifacts
    const tempDir = await fs_1.default.mkdtempSync(`${os_1.default.tmpdir()}/`);
    try {
        // Get directory for mangrove-core module's abis
        const coreDir = path_1.default.parse(require.resolve("@mangrovedao/mangrove-core")).dir;
        const coreAbisDir = `${coreDir}/dist/mangrove-abis`;
        // Get directory for local abis
        const localAbisDir = path_1.default.join(process.cwd(), "src/constants/artifacts");
        if (!fs_1.default.existsSync(localAbisDir)) {
            throw new Error(`Wrong artifacts directory path ${localAbisDir}`);
        }
        // copy all inputs to temp dir. later dirs will have precedence.
        for (const inDir of [coreAbisDir, localAbisDir]) {
            shelljs_1.default.cp(`${inDir}/*.json`, tempDir);
        }
        const artifacts = (0, typechain_1.glob)(process.cwd(), [`${tempDir}/*.json`]);
        const res = await (0, typechain_1.runTypeChain)({
            cwd: process.cwd(),
            filesToProcess: artifacts,
            allFiles: artifacts,
            outDir: outDir,
            target: "ethers-v5",
        });
        console.log("runTypechain.ts:", res);
    }
    finally {
        shelljs_1.default.rm("-rf", tempDir);
    }
}
main()
    .catch(console.error)
    .then((result) => {
    process.exit(0);
});
