// @mangrovedao/mangrove.js/src/types/typechain/factories/AaveV3Module__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type { AaveV3Module, AaveV3ModuleInterface } from "../AaveV3Module";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "_addressesProvider",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "_referralCode",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "_interestRateMode",
        type: "uint256",
      },
    ],
    stateMutability: "nonpayable",
    type: "constructor",
  },
  {
    inputs: [],
    name: "IMPLEMENTATION",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "INTEREST_RATE_MODE",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "ORACLE",
    outputs: [
      {
        internalType: "contract IPriceOracleGetter",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "POOL",
    outputs: [
      {
        internalType: "contract IPool",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "REFERRAL_CODE",
    outputs: [
      {
        internalType: "uint16",
        name: "",
        type: "uint16",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
    ],
    name: "_staticdelegatecall",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "underlying",
        type: "address",
      },
      {
        internalType: "address",
        name: "account",
        type: "address",
      },
    ],
    name: "borrowed",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "contract IERC20",
        name: "asset",
        type: "address",
      },
    ],
    name: "debtToken",
    outputs: [
      {
        internalType: "contract ICreditDelegationToken",
        name: "debtTkn",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "contract IERC20",
        name: "asset",
        type: "address",
      },
      {
        internalType: "bool",
        name: "tryBorrow",
        type: "bool",
      },
      {
        internalType: "address",
        name: "onBehalf",
        type: "address",
      },
    ],
    name: "maxGettableUnderlying",
    outputs: [
      {
        internalType: "uint256",
        name: "maxRedeemableUnderlying",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxBorrowAfterRedeemInUnderlying",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "contract IERC20",
        name: "asset",
        type: "address",
      },
    ],
    name: "overlying",
    outputs: [
      {
        internalType: "contract IERC20",
        name: "aToken",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

export class AaveV3Module__factory {
  static readonly abi = _abi;
  static createInterface(): AaveV3ModuleInterface {
    return new utils.Interface(_abi) as AaveV3ModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AaveV3Module {
    return new Contract(address, _abi, signerOrProvider) as AaveV3Module;
  }
}
