// @mangrovedao/mangrove.js/src/types/typechain/factories/MgvReader__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type { MgvReader, MgvReaderInterface } from "../MgvReader";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "mgv",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "constructor",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "outVolume",
        type: "uint256",
      },
    ],
    name: "getFee",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "gasreq",
        type: "uint256",
      },
    ],
    name: "getProvision",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "ofr_gasreq",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "ofr_gasprice",
        type: "uint256",
      },
    ],
    name: "getProvision",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "global",
    outputs: [
      {
        internalType: "GlobalPacked",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "globalUnpacked",
    outputs: [
      {
        components: [
          {
            internalType: "address",
            name: "monitor",
            type: "address",
          },
          {
            internalType: "bool",
            name: "useOracle",
            type: "bool",
          },
          {
            internalType: "bool",
            name: "notify",
            type: "bool",
          },
          {
            internalType: "uint256",
            name: "gasprice",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "gasmax",
            type: "uint256",
          },
          {
            internalType: "bool",
            name: "dead",
            type: "bool",
          },
        ],
        internalType: "struct GlobalUnpacked",
        name: "",
        type: "tuple",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
    ],
    name: "isEmptyOB",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "tkn0",
        type: "address",
      },
      {
        internalType: "address",
        name: "tkn1",
        type: "address",
      },
    ],
    name: "isMarketOpen",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
    ],
    name: "local",
    outputs: [
      {
        internalType: "LocalPacked",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
    ],
    name: "localUnpacked",
    outputs: [
      {
        components: [
          {
            internalType: "bool",
            name: "active",
            type: "bool",
          },
          {
            internalType: "uint256",
            name: "fee",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "density",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "offer_gasbase",
            type: "uint256",
          },
          {
            internalType: "bool",
            name: "lock",
            type: "bool",
          },
          {
            internalType: "uint256",
            name: "best",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "last",
            type: "uint256",
          },
        ],
        internalType: "struct LocalUnpacked",
        name: "",
        type: "tuple",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "tkn0",
        type: "address",
      },
      {
        internalType: "address",
        name: "tkn1",
        type: "address",
      },
    ],
    name: "marketConfig",
    outputs: [
      {
        components: [
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config01",
            type: "tuple",
          },
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config10",
            type: "tuple",
          },
        ],
        internalType: "struct MgvReader.MarketConfig",
        name: "config",
        type: "tuple",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "takerWants",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "takerGives",
        type: "uint256",
      },
      {
        internalType: "bool",
        name: "fillWants",
        type: "bool",
      },
      {
        internalType: "bool",
        name: "accumulate",
        type: "bool",
      },
    ],
    name: "marketOrder",
    outputs: [
      {
        components: [
          {
            internalType: "uint256",
            name: "totalGot",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalGave",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalGasreq",
            type: "uint256",
          },
        ],
        internalType: "struct VolumeData[]",
        name: "",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "takerWants",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "takerGives",
        type: "uint256",
      },
      {
        internalType: "bool",
        name: "fillWants",
        type: "bool",
      },
    ],
    name: "marketOrder",
    outputs: [
      {
        components: [
          {
            internalType: "uint256",
            name: "totalGot",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalGave",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "totalGasreq",
            type: "uint256",
          },
        ],
        internalType: "struct VolumeData[]",
        name: "",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "gasreq",
        type: "uint256",
      },
    ],
    name: "minVolume",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "outVolume",
        type: "uint256",
      },
    ],
    name: "minusFee",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "numOpenMarkets",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "fromId",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxOffers",
        type: "uint256",
      },
    ],
    name: "offerList",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
      {
        internalType: "uint256[]",
        name: "",
        type: "uint256[]",
      },
      {
        components: [
          {
            internalType: "uint256",
            name: "prev",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "next",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "wants",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "gives",
            type: "uint256",
          },
        ],
        internalType: "struct OfferUnpacked[]",
        name: "",
        type: "tuple[]",
      },
      {
        components: [
          {
            internalType: "address",
            name: "maker",
            type: "address",
          },
          {
            internalType: "uint256",
            name: "gasreq",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "offer_gasbase",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "gasprice",
            type: "uint256",
          },
        ],
        internalType: "struct OfferDetailUnpacked[]",
        name: "",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "fromId",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxOffers",
        type: "uint256",
      },
    ],
    name: "offerListEndPoints",
    outputs: [
      {
        internalType: "uint256",
        name: "startId",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "length",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "from",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxLen",
        type: "uint256",
      },
      {
        internalType: "bool",
        name: "withConfig",
        type: "bool",
      },
    ],
    name: "openMarkets",
    outputs: [
      {
        internalType: "address[2][]",
        name: "markets",
        type: "address[2][]",
      },
      {
        components: [
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config01",
            type: "tuple",
          },
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config10",
            type: "tuple",
          },
        ],
        internalType: "struct MgvReader.MarketConfig[]",
        name: "configs",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bool",
        name: "withConfig",
        type: "bool",
      },
    ],
    name: "openMarkets",
    outputs: [
      {
        internalType: "address[2][]",
        name: "",
        type: "address[2][]",
      },
      {
        components: [
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config01",
            type: "tuple",
          },
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config10",
            type: "tuple",
          },
        ],
        internalType: "struct MgvReader.MarketConfig[]",
        name: "",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "from",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxLen",
        type: "uint256",
      },
    ],
    name: "openMarkets",
    outputs: [
      {
        internalType: "address[2][]",
        name: "markets",
        type: "address[2][]",
      },
      {
        components: [
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config01",
            type: "tuple",
          },
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config10",
            type: "tuple",
          },
        ],
        internalType: "struct MgvReader.MarketConfig[]",
        name: "configs",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "openMarkets",
    outputs: [
      {
        internalType: "address[2][]",
        name: "",
        type: "address[2][]",
      },
      {
        components: [
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config01",
            type: "tuple",
          },
          {
            components: [
              {
                internalType: "bool",
                name: "active",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "fee",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "density",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "offer_gasbase",
                type: "uint256",
              },
              {
                internalType: "bool",
                name: "lock",
                type: "bool",
              },
              {
                internalType: "uint256",
                name: "best",
                type: "uint256",
              },
              {
                internalType: "uint256",
                name: "last",
                type: "uint256",
              },
            ],
            internalType: "struct LocalUnpacked",
            name: "config10",
            type: "tuple",
          },
        ],
        internalType: "struct MgvReader.MarketConfig[]",
        name: "",
        type: "tuple[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "tkn0",
        type: "address",
      },
      {
        internalType: "address",
        name: "tkn1",
        type: "address",
      },
    ],
    name: "order",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "pure",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "outbound_tkn",
        type: "address",
      },
      {
        internalType: "address",
        name: "inbound_tkn",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "fromId",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxOffers",
        type: "uint256",
      },
    ],
    name: "packedOfferList",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
      {
        internalType: "uint256[]",
        name: "",
        type: "uint256[]",
      },
      {
        internalType: "OfferPacked[]",
        name: "",
        type: "uint256[]",
      },
      {
        internalType: "OfferDetailPacked[]",
        name: "",
        type: "uint256[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "tkn0",
        type: "address",
      },
      {
        internalType: "address",
        name: "tkn1",
        type: "address",
      },
    ],
    name: "updateMarket",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class MgvReader__factory {
  static readonly abi = _abi;
  static createInterface(): MgvReaderInterface {
    return new utils.Interface(_abi) as MgvReaderInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): MgvReader {
    return new Contract(address, _abi, signerOrProvider) as MgvReader;
  }
}
