// @mangrovedao/mangrove.js/dist/nodejs/types/typechain/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestToken__factory = exports.SimpleTestMaker__factory = exports.OfferMaker__factory = exports.Multicall2__factory = exports.MgvReader__factory = exports.MgvOracle__factory = exports.MgvCleaner__factory = exports.MangroveOrderEnriched__factory = exports.MangroveOrder__factory = exports.Mangrove__factory = exports.ILiquidityProvider__factory = exports.IERC20__factory = exports.ICreditDelegationToken__factory = exports.AccessControlled__factory = exports.AbstractRouter__factory = exports.AaveV3Module__factory = exports.AaveDeepRouter__factory = exports.factories = void 0;
exports.factories = __importStar(require("./factories"));
var AaveDeepRouter__factory_1 = require("./factories/AaveDeepRouter__factory");
Object.defineProperty(exports, "AaveDeepRouter__factory", { enumerable: true, get: function () { return AaveDeepRouter__factory_1.AaveDeepRouter__factory; } });
var AaveV3Module__factory_1 = require("./factories/AaveV3Module__factory");
Object.defineProperty(exports, "AaveV3Module__factory", { enumerable: true, get: function () { return AaveV3Module__factory_1.AaveV3Module__factory; } });
var AbstractRouter__factory_1 = require("./factories/AbstractRouter__factory");
Object.defineProperty(exports, "AbstractRouter__factory", { enumerable: true, get: function () { return AbstractRouter__factory_1.AbstractRouter__factory; } });
var AccessControlled__factory_1 = require("./factories/AccessControlled__factory");
Object.defineProperty(exports, "AccessControlled__factory", { enumerable: true, get: function () { return AccessControlled__factory_1.AccessControlled__factory; } });
var ICreditDelegationToken__factory_1 = require("./factories/ICreditDelegationToken__factory");
Object.defineProperty(exports, "ICreditDelegationToken__factory", { enumerable: true, get: function () { return ICreditDelegationToken__factory_1.ICreditDelegationToken__factory; } });
var IERC20__factory_1 = require("./factories/IERC20__factory");
Object.defineProperty(exports, "IERC20__factory", { enumerable: true, get: function () { return IERC20__factory_1.IERC20__factory; } });
var ILiquidityProvider__factory_1 = require("./factories/ILiquidityProvider__factory");
Object.defineProperty(exports, "ILiquidityProvider__factory", { enumerable: true, get: function () { return ILiquidityProvider__factory_1.ILiquidityProvider__factory; } });
var Mangrove__factory_1 = require("./factories/Mangrove__factory");
Object.defineProperty(exports, "Mangrove__factory", { enumerable: true, get: function () { return Mangrove__factory_1.Mangrove__factory; } });
var MangroveOrder__factory_1 = require("./factories/MangroveOrder__factory");
Object.defineProperty(exports, "MangroveOrder__factory", { enumerable: true, get: function () { return MangroveOrder__factory_1.MangroveOrder__factory; } });
var MangroveOrderEnriched__factory_1 = require("./factories/MangroveOrderEnriched__factory");
Object.defineProperty(exports, "MangroveOrderEnriched__factory", { enumerable: true, get: function () { return MangroveOrderEnriched__factory_1.MangroveOrderEnriched__factory; } });
var MgvCleaner__factory_1 = require("./factories/MgvCleaner__factory");
Object.defineProperty(exports, "MgvCleaner__factory", { enumerable: true, get: function () { return MgvCleaner__factory_1.MgvCleaner__factory; } });
var MgvOracle__factory_1 = require("./factories/MgvOracle__factory");
Object.defineProperty(exports, "MgvOracle__factory", { enumerable: true, get: function () { return MgvOracle__factory_1.MgvOracle__factory; } });
var MgvReader__factory_1 = require("./factories/MgvReader__factory");
Object.defineProperty(exports, "MgvReader__factory", { enumerable: true, get: function () { return MgvReader__factory_1.MgvReader__factory; } });
var Multicall2__factory_1 = require("./factories/Multicall2__factory");
Object.defineProperty(exports, "Multicall2__factory", { enumerable: true, get: function () { return Multicall2__factory_1.Multicall2__factory; } });
var OfferMaker__factory_1 = require("./factories/OfferMaker__factory");
Object.defineProperty(exports, "OfferMaker__factory", { enumerable: true, get: function () { return OfferMaker__factory_1.OfferMaker__factory; } });
var SimpleTestMaker__factory_1 = require("./factories/SimpleTestMaker__factory");
Object.defineProperty(exports, "SimpleTestMaker__factory", { enumerable: true, get: function () { return SimpleTestMaker__factory_1.SimpleTestMaker__factory; } });
var TestToken__factory_1 = require("./factories/TestToken__factory");
Object.defineProperty(exports, "TestToken__factory", { enumerable: true, get: function () { return TestToken__factory_1.TestToken__factory; } });
