// @mangrovedao/mangrove.js/dist/nodejs/types/typechain/factories/MgvCleaner__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MgvCleaner__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "mgv",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "outbound_tkn",
                type: "address",
            },
            {
                internalType: "address",
                name: "inbound_tkn",
                type: "address",
            },
            {
                internalType: "uint256[4][]",
                name: "targets",
                type: "uint256[4][]",
            },
            {
                internalType: "bool",
                name: "fillWants",
                type: "bool",
            },
        ],
        name: "collect",
        outputs: [
            {
                internalType: "uint256",
                name: "bal",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "outbound_tkn",
                type: "address",
            },
            {
                internalType: "address",
                name: "inbound_tkn",
                type: "address",
            },
            {
                internalType: "uint256[4][]",
                name: "targets",
                type: "uint256[4][]",
            },
            {
                internalType: "bool",
                name: "fillWants",
                type: "bool",
            },
            {
                internalType: "address",
                name: "takerToImpersonate",
                type: "address",
            },
        ],
        name: "collectByImpersonation",
        outputs: [
            {
                internalType: "uint256",
                name: "bal",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        stateMutability: "payable",
        type: "receive",
    },
];
class MgvCleaner__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MgvCleaner__factory = MgvCleaner__factory;
MgvCleaner__factory.abi = _abi;
