// @mangrovedao/mangrove.js/dist/nodejs/types/typechain/factories/MgvOracle__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MgvOracle__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "governance_",
                type: "address",
            },
            {
                internalType: "address",
                name: "initialMutator_",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "density",
                type: "uint256",
            },
        ],
        name: "SetDensity",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "gasPrice",
                type: "uint256",
            },
        ],
        name: "SetGasprice",
        type: "event",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "outbound_tkn",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "inbound_tkn",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "offerId",
                        type: "uint256",
                    },
                    {
                        internalType: "OfferPacked",
                        name: "offer",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "wants",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "gives",
                        type: "uint256",
                    },
                    {
                        internalType: "OfferDetailPacked",
                        name: "offerDetail",
                        type: "uint256",
                    },
                    {
                        internalType: "GlobalPacked",
                        name: "global",
                        type: "uint256",
                    },
                    {
                        internalType: "LocalPacked",
                        name: "local",
                        type: "uint256",
                    },
                ],
                internalType: "struct MgvLib.SingleOrder",
                name: "sor",
                type: "tuple",
            },
            {
                internalType: "address",
                name: "taker",
                type: "address",
            },
        ],
        name: "notifyFail",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "outbound_tkn",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "inbound_tkn",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "offerId",
                        type: "uint256",
                    },
                    {
                        internalType: "OfferPacked",
                        name: "offer",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "wants",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "gives",
                        type: "uint256",
                    },
                    {
                        internalType: "OfferDetailPacked",
                        name: "offerDetail",
                        type: "uint256",
                    },
                    {
                        internalType: "GlobalPacked",
                        name: "global",
                        type: "uint256",
                    },
                    {
                        internalType: "LocalPacked",
                        name: "local",
                        type: "uint256",
                    },
                ],
                internalType: "struct MgvLib.SingleOrder",
                name: "sor",
                type: "tuple",
            },
            {
                internalType: "address",
                name: "taker",
                type: "address",
            },
        ],
        name: "notifySuccess",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        name: "read",
        outputs: [
            {
                internalType: "uint256",
                name: "gasprice",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "density",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "density",
                type: "uint256",
            },
        ],
        name: "setDensity",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "gasPrice",
                type: "uint256",
            },
        ],
        name: "setGasPrice",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "mutator_",
                type: "address",
            },
        ],
        name: "setMutator",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class MgvOracle__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MgvOracle__factory = MgvOracle__factory;
MgvOracle__factory.abi = _abi;
